pyface/base_application.py

```diff
@@ -0,0 +1,395 @@
+        """ Exits the application.
+
+        This method handles a request to shut down the application by the user,
+        eg. from a menu.  If force is False, the application can potentially
+        veto the close event, leaving the application in the state that it was
+        before the exit method was called.
+
+        Parameters
+        ----------
+        force : bool, optional (default False)
+            If set, windows will receive no closing events and will be
+            destroyed unconditionally. This can be useful for reliably tearing
+            down regression tests, but should be used with caution.
+
+        Raises
+        ------
+        ApplicationExit
+            Some subclasses may trigger the exit by raising ApplicationExit.
+        """
+        logger.info('---- Application exit started ----')
+        if force or self._can_exit():
+            try:
+                self._prepare_exit()
+            except Exception:
+                logger.exception("Error preparing for application exit")
+            finally:
+                logger.info('---- Application exit ----')
+                self._exit()
+        else:
+            logger.info('---- Application exit vetoed ----')
+
+    # Initialization utilities -----------------------------------------------
+
+    def setup_logging(self):
+        """ Initialize logger.
+
+        Default behaviour is to log to a rotating file handler whose name is
+        the id of the app.
+        """
+        root_logger = logging.getLogger()
+
+        if DEBUG:
+            handler = logging.StreamHandler()
+            root_logger.setLevel(logging.DEBUG)
+        else:
+            #: set up a rotating file handler
+            log_dir = os.path.join(self.home, 'log')
+            log_file = os.path.join(log_dir, self.id + '.log')
+            if not os.path.exists(log_dir):
+                os.makedirs(log_dir)
+            handler = RotatingFileHandler(
+                log_file, LOG_FILE_MAX_SIZE, LOG_FILE_BACKUPS)
+
+        root_logger.addHandler(handler)
+        self.logging_handlers.append(handler)
+
+    def initialize_application_home(self):
+        """ Set up the home directory for the application
+
+        This is where logs, preference files and other config files will be
+        stored.
+        """
+        if not os.path.exists(self.home):
+            logger.info('Application home directory does not exist, creating')
+            os.makedirs(self.home)
+
+    def install_excepthook(self):
+        """ Install an exception hook to catch unhandled exceptions
+
+        This permits applications to do things like catch and display an error
+        dialog when something unexpectedly goes wrong.
+        """
+        self._old_excepthook = sys.excepthook
+        sys.excepthook = self._excepthook
+        logger.debug('Exception hook installed')
+
+    # Teardown utilities -----------------------------------------------
+
+    def reset_logging(self):
```

Comment:
Same comment as above about `setup_logging`: I don't think it is this class' responsibility to manager the root logger, and it should be left to user code if they want their application object to implement this.

Reply:
Moved to the example (though not keeping the handles on the handlers).